File: bindings/android/scion-http3-android/src/main/kotlin/com/anapaya/scion/http3/internal/Mapping.kt
// Copyright 2026 Anapaya Systems
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package com.anapaya.scion.http3.internal

import com.anapaya.scion.http3.PreferredUnderlay
import com.anapaya.scion.http3.ScionHttp3Exception
import com.anapaya.scion.http3.ScionHttp3Headers
import com.anapaya.scion.http3.ScionHttp3Request
import com.anapaya.scion.http3.ScionHttp3Response
import com.anapaya.scion.http3.ScionHttp3ResponseBody
import com.anapaya.scion.http3.SnapConfig
import com.anapaya.scion.http3.TrustAnchors
import com.anapaya.scion.http3.UdpConfig
import com.anapaya.scion.http3.uniffi.ClientConfig
import com.anapaya.scion.http3.uniffi.Header
import com.anapaya.scion.http3.uniffi.HttpRequest
import com.anapaya.scion.http3.uniffi.HttpResponse
import com.anapaya.scion.http3.uniffi.Underlay
import com.anapaya.scion.http3.uniffi.ScionHttp3Exception as FfiException
import com.anapaya.scion.http3.uniffi.SnapConfig as FfiSnapConfig
import com.anapaya.scion.http3.uniffi.TimeoutPhase as FfiTimeoutPhase
import com.anapaya.scion.http3.uniffi.TrustAnchors as FfiTrustAnchors
import com.anapaya.scion.http3.uniffi.UdpConfig as FfiUdpConfig

// The one place the public types and the generated ones meet.
//
// Confining it to a single file is what keeps the two `ScionHttp3Client` and `TrustAnchors` types,
// one public and one generated, from being a standing source of confusion: this is the only file
// that imports both, and it aliases the generated side.

/** The request as the stack receives it. */
internal fun ScionHttp3Request.toFfi(): HttpRequest =
    HttpRequest(
        url = url,
        method = method,
        headers = headers.map { Header(name = it.name, value = it.value) },
        // Absent and empty are different below: absent sends no body at all, where an empty one
        // sends a body of zero bytes. Preserve which the caller chose.
        body = body?.bytesNoCopy(),
        targets = targets.map { it.toString() },
        requestTimeoutMs = requestTimeoutMillis?.toULong(),
        maxResponseBodyBytes = maxResponseBodyBytes?.toULong(),
    )

/** The response as the application sees it. */
internal fun HttpResponse.toPublic(request: ScionHttp3Request): ScionHttp3Response =
    ScionHttp3Response(
        code = status.toInt(),
        headers = headers.toPublic(),
        body = ScionHttp3ResponseBody(body),
        // The stack reports no trailer section as an empty list, which cannot be told apart from an
        // empty section that was sent. Reported as absent, which is true far more often.
        trailers = if (trailers.isEmpty()) null else trailers.toPublic(),
        request = request,
    )

private fun List<Header>.toPublic(): ScionHttp3Headers =
    ScionHttp3Headers.ofEntries(
        map { ScionHttp3Headers.Entry(name = it.name, value = it.value) },
    )

/**
 * The public exception for a failure from the stack.
 *
 * `isRetryable` and `detail` are carried across untouched. Retryability is the stack's judgement,
 * which depends on what actually failed and not on which arm it lands in, so deriving it here would
 * be a second, worse answer to a question already answered.
 *
 * The `when` is exhaustive over a sealed type, so a new arm below is a compilation error here rather
 * than something that quietly becomes [ScionHttp3Exception.Internal].
 */
internal fun FfiException.toPublic(): ScionHttp3Exception =
    when (this) {
        is FfiException.StackBuild ->
            ScionHttp3Exception.Connectivity(retryable, detail, this)
        is FfiException.Resolution ->
            ScionHttp3Exception.Resolution(host, retryable, detail, this)
        is FfiException.Connect ->
            ScionHttp3Exception.Connect(host, port.toInt(), retryable, detail, this)
        is FfiException.Tls ->
            ScionHttp3Exception.Tls(host, retryable, detail, this)
        is FfiException.StreamReset ->
            ScionHttp3Exception.StreamReset(code.toLong(), retryable, detail, this)
        is FfiException.Protocol ->
            ScionHttp3Exception.Protocol(retryable, detail, this)
        is FfiException.ConnectionLimit ->
            ScionHttp3Exception.ConnectionLimit(retryable, detail, this)
        is FfiException.BodyTooLarge ->
            ScionHttp3Exception.BodyTooLarge(limit.toLong(), retryable, detail, this)
        is FfiException.Timeout ->
            ScionHttp3Exception.Timeout(
                phase.toPublic(),
                timeoutMs.toLong(),
                retryable,
                detail,
                this,
            )
        is FfiException.InvalidRequest ->
            ScionHttp3Exception.InvalidRequest(retryable, detail, this)
        is FfiException.Closed ->
            ScionHttp3Exception.Closed(retryable, detail, this)
        is FfiException.Internal ->
            ScionHttp3Exception.Internal(retryable, detail, this)
    }

private fun FfiTimeoutPhase.toPublic(): ScionHttp3Exception.TimeoutPhase =
    when (this) {
        FfiTimeoutPhase.CONNECT -> ScionHttp3Exception.TimeoutPhase.CONNECT
        FfiTimeoutPhase.REQUEST -> ScionHttp3Exception.TimeoutPhase.REQUEST
        FfiTimeoutPhase.BODY -> ScionHttp3Exception.TimeoutPhase.BODY
        FfiTimeoutPhase.OTHER -> ScionHttp3Exception.TimeoutPhase.OTHER
    }

/**
 * Applies the configured settings on top of the stack's own defaults.
 *
 * [base] is what `defaultClientConfig()` returned, so every field this leaves alone keeps the value
 * the SCION stack chose. That is the point: the defaults have exactly one home, and it is not here.
 */
internal fun ClientSettings.applyTo(
    base: ClientConfig,
    trust: FfiTrustAnchors,
): ClientConfig =
    base.copy(
        endhostApiUrl = endhostApiUrl,
        authToken = authToken ?: base.authToken,
        preferredUnderlay = preferredUnderlay?.toFfi() ?: base.preferredUnderlay,
        snap = snap?.toFfi() ?: base.snap,
        udp = udp?.toFfi() ?: base.udp,
        trust = trust,
        connectTimeoutMs = connectTimeoutMillis?.toULong() ?: base.connectTimeoutMs,
        requestTimeoutMs = requestTimeoutMillis?.toULong() ?: base.requestTimeoutMs,
        idleConnectionTimeoutMs =
            idleConnectionTimeoutMillis?.toULong() ?: base.idleConnectionTimeoutMs,
        maxOrigins = maxOrigins?.toUInt() ?: base.maxOrigins,
        connectionAttemptDelayMs =
            connectionAttemptDelayMillis?.toULong() ?: base.connectionAttemptDelayMs,
        maxResponseBodyBytes = maxResponseBodyBytes?.toULong() ?: base.maxResponseBodyBytes,
    )

private fun PreferredUnderlay.toFfi(): Underlay =
    when (this) {
        PreferredUnderlay.SNAP -> Underlay.SNAP
        PreferredUnderlay.UDP -> Underlay.UDP
    }

private fun SnapConfig.toFfi(): FfiSnapConfig =
    FfiSnapConfig(
        dpIndex = dataPlaneIndex?.toUInt(),
        staticIdentity = staticIdentity,
    )

private fun UdpConfig.toFfi(): FfiUdpConfig =
    FfiUdpConfig(
        outboundIps = outboundIps,
        nextHopResolverFetchIntervalMs = nextHopResolverFetchIntervalMillis?.toULong(),
    )

/**
 * The trust anchors the stack should use.
 *
 * The platform's anchors are read here and passed down as bytes, because the stack's TLS
 * implementation cannot read the keystore they live in. Its own "system default" would configure
 * nothing at all on Android, so it is never what this returns.
 */
internal fun TrustAnchors.toFfi(store: SystemTrustStore): FfiTrustAnchors =
    when (this) {
        is TrustAnchors.PlatformAnchors -> FfiTrustAnchors.Pem(store.anchorsPem())
        is TrustAnchors.PinnedAnchors -> FfiTrustAnchors.Pem(pem())
        is TrustAnchors.NoVerification -> FfiTrustAnchors.InsecureNoVerify
    }

File: bindings/apple/scion-http3-swift/Sources/ScionHTTP3/Internal/Mapping.swift
// Copyright 2026 Anapaya Systems
import Foundation
internal import ScionHTTP3Uniffi

typealias FfiClient = ScionHTTP3Uniffi.ScionHttp3Client
typealias FfiError = ScionHTTP3Uniffi.ScionHttp3Error
typealias FfiTrustAnchors = ScionHTTP3Uniffi.TrustAnchors
typealias FfiSnapConfig = ScionHTTP3Uniffi.SnapConfig
typealias FfiUdpConfig = ScionHTTP3Uniffi.UdpConfig
typealias FfiTimeoutPhase = ScionHTTP3Uniffi.TimeoutPhase

/// The request as the stack receives it.
///
/// Throws `ScionHttp3Error.invalidRequest` for what the request itself gets wrong, before anything
/// is sent.
func ffiRequest(from request: ScionHttp3Request) throws -> HttpRequest {
    try request.validate()
    var headers = request.headers
    if let contentType = request.body?.contentType {
        headers.addIfAbsent("content-type", contentType)
    }
    return HttpRequest(
        url: request.url,
        method: request.method.rawValue,
        headers: headers.map { Header(name: $0.name, value: $0.value) },
        // Absent and empty are different below: absent sends no body at all, where an empty one
        // sends a body of zero bytes. Preserve which the caller chose.
        body: request.body?.data,
        targets: request.targets.map(\.description),
        requestTimeoutMs: request.requestTimeout.map(milliseconds),
        maxResponseBodyBytes: request.maxResponseBodyBytes.map { UInt64(clamping: $0) })
}

/// The response as the application sees it.
func publicResponse(
    from response: HttpResponse, to request: ScionHttp3Request, keepAlive: any Sendable
) -> ScionHttp3Response {
    ScionHttp3Response(
        code: Int(response.status),
        headers: publicHeaders(response.headers),
        body: ScionHttp3ResponseBody(response.body, keepAlive: keepAlive),
        trailers: response.trailers.isEmpty ? nil : publicHeaders(response.trailers),
        request: request)
}

private func publicHeaders(_ headers: [Header]) -> ScionHttp3Headers {
    ScionHttp3Headers(
        entries: headers.map { ScionHttp3Headers.Entry(name: $0.name, value: $0.value) })
}

/// The public error for a failure from the stack; any other error passes through unchanged.
func publicError(_ error: any Error) -> any Error {
    guard let ffi = error as? FfiError else { return error }
    switch ffi {
    case .StackBuild(let retryable, let detail):
        return ScionHttp3Error.connectivity(detail: detail, retryable: retryable)
    case .Resolution(let host, let retryable, let detail):
        return ScionHttp3Error.resolution(host: host, detail: detail, retryable: retryable)
    case .Connect(let host, let port, let retryable, let detail):
        return ScionHttp3Error.connect(
            host: host, port: Int(port), detail: detail, retryable: retryable)
    case .Tls(let host, let retryable, let detail):
        return ScionHttp3Error.tls(host: host, detail: detail, retryable: retryable)
    case .StreamReset(let code, let retryable, let detail):
        return ScionHttp3Error.streamReset(code: code, detail: detail, retryable: retryable)
    case .Protocol(let retryable, let detail):
        return ScionHttp3Error.protocolViolation(detail: detail, retryable: retryable)
    case .ConnectionLimit(let retryable, let detail):
        return ScionHttp3Error.connectionLimit(detail: detail, retryable: retryable)
    case .BodyTooLarge(let limit, let retryable, let detail):
        return ScionHttp3Error.bodyTooLarge(
            limit: Int(clamping: limit), detail: detail, retryable: retryable)
    case .Timeout(let phase, let timeoutMs, let retryable, let detail):
        return ScionHttp3Error.timeout(
            phase: publicPhase(phase), after: TimeInterval(timeoutMs) / 1000, detail: detail,
            retryable: retryable)
    case .InvalidRequest(_, let detail):
        return ScionHttp3Error.invalidRequest(detail: detail)
    case .Closed:
        return ScionHttp3Error.closed
    case .Cancelled:
        return CancellationError()
    case .Internal(_, let detail):
        return ScionHttp3Error.internalError(detail: detail)
    }
}

private func publicPhase(_ phase: FfiTimeoutPhase) -> ScionHttp3Error.TimeoutPhase {
    switch phase {
    case .connect: return .connect
    case .request: return .request
    case .body: return .body
    case .other: return .other
    }
}

/// Whole milliseconds, as the FFI carries every duration.
func milliseconds(_ seconds: TimeInterval) -> UInt64 {
    UInt64(clamping: Int64((seconds * 1000).rounded()))
}

extension ClientSettings {
    /// Applies the configured settings on top of the stack's own defaults.
    func applyTo(_ base: ClientConfig) -> ClientConfig {
        var config = base
        config.endhostApiUrl = endhostApiUrl
        config.authToken = authToken ?? base.authToken
        config.preferredUnderlay = preferredUnderlay.map(ffiUnderlay) ?? base.preferredUnderlay
        config.snap = FfiSnapConfig(
            dpIndex: snap.dataPlaneIndex.map { UInt32(clamping: $0) } ?? base.snap.dpIndex,
            staticIdentity: snap.staticIdentity ?? base.snap.staticIdentity)
        config.udp = FfiUdpConfig(
            outboundIps: udp.outboundIps.isEmpty ? base.udp.outboundIps : udp.outboundIps,
            nextHopResolverFetchIntervalMs: udp.nextHopResolverFetchInterval.map(milliseconds)
                ?? base.udp.nextHopResolverFetchIntervalMs)
        config.trust = ffiTrustAnchors(trust)
        config.connectTimeoutMs = connectTimeout.map(milliseconds) ?? base.connectTimeoutMs
        config.requestTimeoutMs = requestTimeout.map(milliseconds) ?? base.requestTimeoutMs
        config.idleConnectionTimeoutMs =
            idleConnectionTimeout.map(milliseconds) ?? base.idleConnectionTimeoutMs
        config.maxOrigins = maxOrigins.map { UInt32(clamping: $0) } ?? base.maxOrigins
        config.connectionAttemptDelayMs =
            connectionAttemptDelay.map(milliseconds) ?? base.connectionAttemptDelayMs
        config.maxResponseBodyBytes =
            maxResponseBodyBytes.map { UInt64(clamping: $0) } ?? base.maxResponseBodyBytes
        return config
    }
}

private func ffiUnderlay(_ underlay: PreferredUnderlay) -> Underlay {
    switch underlay {
    case .snap: return .snap
    case .udp: return .udp
    }
}

/// The trust anchors the stack should use.
func ffiTrustAnchors(_ trust: TrustAnchors) -> FfiTrustAnchors {
    switch trust.kind {
    case .systemDefault: return .systemDefault
    case .pinned(let pem): return .pem(pem: pem)
    case .insecureNoVerify: return .insecureNoVerify
    }
}
